src/dft/backends/mklgpu/forward.cpp

```diff
@@ -0,0 +1,193 @@
+ONEMKL_EXPORT void compute_forward(descriptor_type &desc, sycl::buffer<input_type, 1> &in,
+                                   sycl::buffer<output_type, 1> &out) {
+    if constexpr (!std::is_same_v<input_type, output_type>) {
+        throw mkl::unimplemented(
+            "DFT", "compute_forward",
+            "MKLGPU does not support out-of-place FFT with different input and output types.");
+    }
+    else {
+        detail::expect_config<dft::detail::config_param::PLACEMENT,
+                              dft::detail::config_value::NOT_INPLACE>(
+            desc, "Unexpected value for placement");
+        return detail::compute_forward(desc, in, out);
+    }
+}
+
+//Out-of-place transform, using config_param::COMPLEX_STORAGE=config_value::REAL_REAL data format
+template <typename descriptor_type, typename input_type, typename output_type>
+ONEMKL_EXPORT void compute_forward(descriptor_type &desc, sycl::buffer<input_type, 1> &in_re,
+                                   sycl::buffer<input_type, 1> &in_im,
+                                   sycl::buffer<output_type, 1> &out_re,
+                                   sycl::buffer<output_type, 1> &out_im) {
+    detail::expect_config<dft::detail::config_param::COMPLEX_STORAGE,
+                          dft::detail::config_value::REAL_REAL>(
+        desc, "Unexpected value for complex storage");
+    throw oneapi::mkl::unimplemented(
+        "DFT", "compute_forward(desc, in_re, in_im, out_re, out_im)",
+        "MKLGPU does not support out-of-place FFT with real-real complex storage.");
+}
+
+//USM version
+
+//In-place transform
+template <typename descriptor_type, typename data_type>
+ONEMKL_EXPORT sycl::event compute_forward(descriptor_type &desc, data_type *inout,
+                                          const std::vector<sycl::event> &dependencies) {
+    detail::expect_config<dft::detail::config_param::PLACEMENT, dft::detail::config_value::INPLACE>(
+        desc, "Unexpected value for placement");
+    return detail::compute_forward(desc, inout, dependencies);
+}
+
+//In-place transform, using config_param::COMPLEX_STORAGE=config_value::REAL_REAL data format
+template <typename descriptor_type, typename data_type>
+ONEMKL_EXPORT sycl::event compute_forward(descriptor_type &desc, data_type *inout_re,
+                                          data_type *inout_im,
+                                          const std::vector<sycl::event> &dependencies) {
+    throw mkl::unimplemented("DFT", "compute_forward",
+                             "MKLGPU does not support compute_forward(desc, inout_re, inout_im).");
+}
+
+//Out-of-place transform
+template <typename descriptor_type, typename input_type, typename output_type>
+ONEMKL_EXPORT sycl::event compute_forward(descriptor_type &desc, input_type *in, output_type *out,
+                                          const std::vector<sycl::event> &dependencies) {
+    if constexpr (!std::is_same_v<input_type, output_type>) {
```

Comment:
I believe in intel oneMKL we added support for mixed types, but that might not be refelected in the specs, https://github.com/intel-innersource/libraries.performance.math.mkl/pull/1516

Reply:
I don't have access to that repo but that does indeed seem to be working. I've enabled the forward and backward transforms for both buffers and usm in [92b7fa10937442ea22d7a149cf4afa7312623187](https://github.com/oneapi-src/oneMKL/pull/259/commits/92b7fa10937442ea22d7a149cf4afa7312623187)